cmd/common/runwithspec_test.go

```diff
@@ -0,0 +1,260 @@
+// Copyright 2024 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import (
+	"fmt"
+	"os"
+	"testing"
+
+	"github.com/spf13/cobra"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/environment"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-service/api"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/params"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/runtime"
+	grpcruntime "github.com/inspektor-gadget/inspektor-gadget/pkg/runtime/grpc"
+)
+
+type testMode int
+
+const (
+	testModeBoth testMode = iota
+	testModeInteractiveOnly
+	testModeDetachOnly
+)
+
+type testSpec struct {
+	Name           string
+	Manifest       string
+	Mode           testMode
+	AdditionalArgs []string
+	ExpectedParams api.ParamValues
+	ExpectedName   string
+	ExpectedID     string
+	ExpectedTags   []string
+	ExpectError    bool
+}
+
+type testRuntime struct {
+	grpcruntime.Runtime
+	t        *testing.T
+	testSpec *testSpec
+}
+
+func (r *testRuntime) Init(*params.Params) error {
+	return nil
+}
+
+func (r *testRuntime) Close() error {
+	return nil
+}
+
+func (r *testRuntime) GetGadgetInfo(gadgetCtx runtime.GadgetContext, runtimeParams *params.Params, paramValueMap api.ParamValues) (*api.GadgetInfo, error) {
+	panic("unimplemented")
+}
+
+func (r *testRuntime) RunGadget(gadgetCtx runtime.GadgetContext, runtimeParams *params.Params, paramValueMap api.ParamValues) error {
+	if r.testSpec.ExpectedID != "" {
+		assert.Equal(r.t, r.testSpec.ExpectedID, runtimeParams.Get(grpcruntime.ParamID).AsString())
+	}
+	if r.testSpec.ExpectedName != "" {
+		assert.Equal(r.t, r.testSpec.ExpectedName, runtimeParams.Get(grpcruntime.ParamName).AsString())
+	}
+	if r.testSpec.ExpectedTags != nil {
+		assert.Equal(r.t, r.testSpec.ExpectedTags, runtimeParams.Get(grpcruntime.ParamTags).AsStringSlice())
+	}
+	if r.testSpec.ExpectedParams != nil {
+		for k, v := range r.testSpec.ExpectedParams {
+			val, ok := paramValueMap[k]
+			require.True(r.t, ok, "expected paramValue exists for %q", k)
+			require.True(r.t, ok, "expected paramValue to be %q, got %q", v, val)
```

Comment:
`require.Equal(r.t, v, val)` ? 

Also the require package already provides the "expected .. got" message, so we don't need to add it.